File: hrm_app/lib/core/constants/websocket_events.dart
class WebSocketEvents {
  static const String connection = 'connection';
  static const String disconnect = 'disconnect';
  static const String authenticate = 'authenticate';
  static const String authenticated = 'authenticated';
  static const String authenticationFailed = 'authentication_failed';

  static const String joinRoom = 'join_room';
  static const String leaveRoom = 'leave_room';
  static const String roomJoined = 'room_joined';
  static const String roomLeft = 'room_left';

  static const String userConnected = 'user_connected';
  static const String userDisconnected = 'user_disconnected';
  static const String userStatusChanged = 'user_status_changed';

  static const String employeeCreated = 'employee_created';
  static const String employeeUpdated = 'employee_updated';
  static const String employeeDeleted = 'employee_deleted';
  static const String employeeStatusChanged = 'employee_status_changed';
  static const String employeeProfileUpdated = 'employee_profile_updated';

  static const String leaveRequestCreated = 'leave_request_created';
  static const String leaveRequestApproved = 'leave_request_approved';
  static const String leaveRequestRejected = 'leave_request_rejected';
  static const String leaveRequestCancelled = 'leave_request_cancelled';
  static const String leaveBalanceUpdated = 'leave_balance_updated';
  static const String leaveCalendarUpdated = 'leave_calendar_updated';

  static const String payslipRequestCreated = 'payslip_request_created';
  static const String payslipRequestApproved = 'payslip_request_approved';
  static const String payslipRequestRejected = 'payslip_request_rejected';
  static const String payslipGenerated = 'payslip_generated';
  static const String payrollProcessed = 'payroll_processed';
  static const String salaryUpdated = 'salary_updated';

  static const String messageSent = 'message_sent';
  static const String messageReceived = 'message_received';
  static const String messageRead = 'message_read';
  static const String messageDeleted = 'message_deleted';
  static const String conversationCreated = 'conversation_created';
  static const String typingIndicator = 'typing_indicator';
  static const String typingStopped = 'typing_stopped';

  static const String eodSubmitted = 'eod_submitted';
  static const String eodApproved = 'eod_approved';
  static const String eodRejected = 'eod_rejected';
  static const String eodReminder = 'eod_reminder';
  static const String eodDeadlineApproaching = 'eod_deadline_approaching';

  static const String announcementCreated = 'announcement_created';
  static const String announcementUpdated = 'announcement_updated';
  static const String announcementDeleted = 'announcement_deleted';
  static const String notificationSent = 'notification_sent';
  static const String systemAlert = 'system_alert';

  static const String checkIn = 'check_in';
  static const String checkOut = 'check_out';
  static const String breakStart = 'break_start';
  static const String breakEnd = 'break_end';
  static const String overtimeLogged = 'overtime_logged';
  static const String attendanceUpdated = 'attendance_updated';

  static const String customEvent = 'custom_event';

  static const String error = 'error';
  static const String rateLimitExceeded = 'rate_limit_exceeded';
  static const String invalidData = 'invalid_data';

  static const String systemMaintenance = 'system_maintenance';
  static const String systemUpdate = 'system_update';
  static const String connectionStatus = 'connection_status';
  static const String heartbeat = 'heartbeat';
}

class WebSocketRooms {
  static const String adminRoom = 'admin_room';
  static const String hrRoom = 'hr_room';
  static const String employeeRoom = 'employee_room';
  static const String companyWide = 'company_wide';

  static String userRoom(String userId) => 'user_$userId';
  static String departmentRoom(String deptId) => 'department_$deptId';
  static String projectRoom(String projectId) => 'project_$projectId';
  static String teamRoom(String teamId) => 'team_$teamId';
}

class EventCategories {
  static const String connection = 'connection';
  static const String userManagement = 'user_management';
  static const String employeeManagement = 'employee_management';
  static const String leaveManagement = 'leave_management';
  static const String payroll = 'payroll';
  static const String messaging = 'messaging';
  static const String eod = 'eod';
  static const String announcements = 'announcements';
  static const String attendance = 'attendance';
  static const String system = 'system';
}

class EventPriorities {
  static const String low = 'low';
  static const String medium = 'medium';
  static const String high = 'high';
  static const String critical = 'critical';
}

enum ConnectionStatus {
  disconnected,
  connecting,
  connected,
  reconnecting,
  error,
}

class WebSocketEventData {
  final String event;
  final Map<String, dynamic> data;
  final DateTime timestamp;
  final String? room;
  final String? userId;

  WebSocketEventData({
    required this.event,
    required this.data,
    required this.timestamp,
    this.room,
    this.userId,
  });

  factory WebSocketEventData.fromMap(Map<String, dynamic> map) {
    return WebSocketEventData(
      event: map['event'] ?? '',
      data: Map<String, dynamic>.from(map['data'] ?? {}),
      timestamp: DateTime.tryParse(map['timestamp'] ?? '') ?? DateTime.now(),
      room: map['room'],
      userId: map['userId'],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'event': event,
      'data': data,
      'timestamp': timestamp.toIso8601String(),
      'room': room,
      'userId': userId,
    };
  }
}

File: hrm_backend/src/constants/websocket.events.js
// WebSocket Event Constants
// This file contains all the WebSocket event names used throughout the application

const WEBSOCKET_EVENTS = {
  // Connection Events
  CONNECTION: "connection",
  DISCONNECT: "disconnect",
  AUTHENTICATE: "authenticate",
  AUTHENTICATED: "authenticated",
  AUTHENTICATION_FAILED: "authentication_failed",

  // Room Management Events
  JOIN_ROOM: "join_room",
  LEAVE_ROOM: "leave_room",
  ROOM_JOINED: "room_joined",
  ROOM_LEFT: "room_left",

  // User Events
  USER_CONNECTED: "user_connected",
  USER_DISCONNECTED: "user_disconnected",
  USER_STATUS_CHANGED: "user_status_changed",

  // Employee Management Events
  EMPLOYEE_CREATED: "employee_created",
  EMPLOYEE_UPDATED: "employee_updated",
  EMPLOYEE_DELETED: "employee_deleted",
  EMPLOYEE_STATUS_CHANGED: "employee_status_changed",
  EMPLOYEE_PROFILE_UPDATED: "employee_profile_updated",

  // Leave Management Events
  LEAVE_REQUEST_CREATED: "leave_request_created",
  LEAVE_REQUEST_APPROVED: "leave_request_approved",
  LEAVE_REQUEST_REJECTED: "leave_request_rejected",
  LEAVE_REQUEST_CANCELLED: "leave_request_cancelled",
  LEAVE_BALANCE_UPDATED: "leave_balance_updated",
  LEAVE_CALENDAR_UPDATED: "leave_calendar_updated",

  // Payroll & Payslip Events
  PAYSLIP_REQUEST_CREATED: "payslip_request_created",
  PAYSLIP_REQUEST_APPROVED: "payslip_request_approved",
  PAYSLIP_REQUEST_REJECTED: "payslip_request_rejected",
  PAYSLIP_REQUEST_ON_HOLD: "payslip_request_on_hold",
  PAYSLIP_GENERATED: "payslip_generated",
  PAYROLL_PROCESSED: "payroll_processed",
  SALARY_UPDATED: "salary_updated",

  // Messaging System Events
  MESSAGE_SENT: "message_sent",
  MESSAGE_RECEIVED: "message_received",
  MESSAGE_READ: "message_read",
  // MESSAGE_DELETED and MESSAGE_EDITED removed for organizational security
  CONVERSATION_CREATED: "conversation_created",
  CONVERSATION_UPDATED: "conversation_updated",
  // CONVERSATION_DELETED, CONVERSATION_ARCHIVED, CONVERSATION_UNARCHIVED removed for organizational security
  TYPING_INDICATOR: "typing_indicator",
  TYPING_STOPPED: "typing_stopped",

  // End of Day (EOD) Events
  EOD_SUBMITTED: "eod_submitted",
  EOD_APPROVED: "eod_approved",
  EOD_REJECTED: "eod_rejected",
  EOD_REMINDER: "eod_reminder",
  EOD_DEADLINE_APPROACHING: "eod_deadline_approaching",

  // Announcements & Notifications
  ANNOUNCEMENT_CREATED: "announcement_created",
  ANNOUNCEMENT_UPDATED: "announcement_updated",
  ANNOUNCEMENT_DELETED: "announcement_deleted",
  NOTIFICATION_SENT: "notification_sent",
  SYSTEM_ALERT: "system_alert",

  // Attendance & Time Tracking
  CHECK_IN: "check_in",
  CHECK_OUT: "check_out",
  BREAK_START: "break_start",
  BREAK_END: "break_end",
  OVERTIME_LOGGED: "overtime_logged",
  ATTENDANCE_UPDATED: "attendance_updated",

  // Custom Events
  CUSTOM_EVENT: "custom_event",

  // Error Events
  ERROR: "error",
  RATE_LIMIT_EXCEEDED: "rate_limit_exceeded",
  INVALID_DATA: "invalid_data",

  // System Events
  SYSTEM_MAINTENANCE: "system_maintenance",
  SYSTEM_UPDATE: "system_update",
  CONNECTION_STATUS: "connection_status",
  HEARTBEAT: "heartbeat",
};

// Room Names
const WEBSOCKET_ROOMS = {
  ADMIN_ROOM: "admin_room",
  HR_ROOM: "hr_room",
  EMPLOYEE_ROOM: "employee_room",
  COMPANY_WIDE: "company_wide",
  USER_ROOM: (userId) => `user_${userId}`,
  DEPARTMENT_ROOM: (deptId) => `department_${deptId}`,
  PROJECT_ROOM: (projectId) => `project_${projectId}`,
  TEAM_ROOM: (teamId) => `team_${teamId}`,
};

// Event Categories for better organization
const EVENT_CATEGORIES = {
  CONNECTION: "connection",
  USER_MANAGEMENT: "user_management",
  EMPLOYEE_MANAGEMENT: "employee_management",
  LEAVE_MANAGEMENT: "leave_management",
  PAYROLL: "payroll",
  MESSAGING: "messaging",
  EOD: "eod",
  ANNOUNCEMENTS: "announcements",
  ATTENDANCE: "attendance",
  SYSTEM: "system",
};

// Event Priorities
const EVENT_PRIORITIES = {
  LOW: "low",
  MEDIUM: "medium",
  HIGH: "high",
  CRITICAL: "critical",
};

module.exports = {
  WEBSOCKET_EVENTS,
  WEBSOCKET_ROOMS,
  EVENT_CATEGORIES,
  EVENT_PRIORITIES,
};
